### PR TITLE
SONARJAVA-2885 Rule S2689: Files opened in append mode should not be used with ObjectOutputStream

### PR DESCRIPTION
**Pull request by mpaladin**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/checks/ObjectOutputStreamCheck.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    if (syntaxNode.is(Tree.Kind.NEW_CLASS)) {
+      NewClassTree newClassTree = (NewClassTree) syntaxNode;
+      if (newClassTree.symbolType().is("java.io.ObjectOutputStream")) {
+        ProgramState programState = context.getState();
+        FileOutputStreamConstraint constraint = programState.getConstraint(programState.peekValue(), FileOutputStreamConstraint.class);
+        if (constraint == FileOutputStreamConstraint.APPEND) {
+          context.reportIssue(syntaxNode, this, "This file was opened in append mode.");
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/ObjectOutputStreamCheck.java]: I'm not that much convinced by the message, especially has we are going to have a `FileInputStream` here, not a `File`. No secondaries/flow?

File: java-frontend/src/test/files/se/ObjectOutputStreamCheck.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    FileOutputStream fos = Files.newOutputStream(Paths.get("a"), StandardOpenOption.APPEND);
+    ObjectOutputStream out = new ObjectOutputStream(fos); // Noncompliant
+  }
+  void noncompliant_11() throws IOException {
+    FileOutputStream fos = Files.newOutputStream(Paths.get("a"), StandardOpenOption.DELETE_ON_CLOSE, StandardOpenOption.APPEND);
+    ObjectOutputStream out = new ObjectOutputStream(fos); // Noncompliant
+  }
+  void noncompliant_12() throws IOException {
+    OpenOption openOption = StandardOpenOption.APPEND;
+    FileOutputStream fos = Files.newOutputStream(Paths.get("a"), StandardOpenOption.DELETE_ON_CLOSE, openOption);
```
Comment (m-g-sonar) [inline java-frontend/src/test/files/se/ObjectOutputStreamCheck.java]: :+1:

File: java-frontend/src/test/files/se/ObjectOutputStreamCheck.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  void noncompliant_1(String fileName) throws IOException {
+    FileOutputStream fos = new FileOutputStream(fileName , true);  // fos opened in append mode
+    ObjectOutputStream out = new ObjectOutputStream(fos);  // Noncompliant
+  }
+  void noncompliant_2(String fileName, boolean appendMode) throws IOException {
+    if (!appendMode) return;
+    FileOutputStream fos = new FileOutputStream(fileName, appendMode);  // fos opened in append mode
+    ObjectOutputStream out = new ObjectOutputStream(fos);  // Noncompliant
+  }
+  void noncompliant_3(File file) throws IOException {
```
Comment (m-g-sonar) [inline java-frontend/src/test/files/se/ObjectOutputStreamCheck.java]: You are not importing `java.io.File`, so I'm a bit surprise we are able to resolve this constructor.

File: java-frontend/src/main/java/org/sonar/java/se/checks/ObjectOutputStreamCheck.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    .withAnyParameters();
+
+  public enum FileOutputStreamConstraint implements Constraint {
+    APPEND
+  }
+
+  @Override
+  public ProgramState checkPostStatement(CheckerContext context, Tree syntaxNode) {
+    ProgramState programState = context.getState();
+    if (syntaxNode.is(Tree.Kind.NEW_CLASS)) {
```
[the remark this answers is not in the document]
Comment (mpaladin) [reply java-frontend/src/main/java/org/sonar/java/se/checks/ObjectOutputStreamCheck.java]: @m-g-sonar whether it should be a `switch` or an `if-then-else-if` I don't have a strong opinion, the only argument is performance. However I am against extracting 3 different methods, IMHO it makes the rule implementation more complex in this case, it's pretty simple to read and understand as it is.

File: java-frontend/src/main/java/org/sonar/java/se/checks/ObjectOutputStreamCheck.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+      }
+    } else if (syntaxNode.is(Tree.Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree mit = (MethodInvocationTree) syntaxNode;
+      if (FILES_NEW_OUTPUT_STREAM.matches(mit)) {
+        ProgramState psBeforeInvocation = context.getNode().programState;
+        int optionsNumber = mit.arguments().size() - 1/*skip Path argument*/;
+        for (int i = 0; i < optionsNumber; i++) {
+          FileOutputStreamConstraint argConstraint = psBeforeInvocation.getConstraint(psBeforeInvocation.peekValue(i), FileOutputStreamConstraint.class);
+          if (argConstraint == FileOutputStreamConstraint.APPEND) {
+            programState = programState.addConstraint(programState.peekValue(), FileOutputStreamConstraint.APPEND);
```
[the remark this answers is not in the document]
Comment (mpaladin) [reply java-frontend/src/main/java/org/sonar/java/se/checks/ObjectOutputStreamCheck.java]: @m-g-sonar fair point.

File: java-frontend/src/main/java/org/sonar/java/se/checks/ObjectOutputStreamCheck.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+          }
+        }
+      }
+    } else if (syntaxNode.is(Tree.Kind.MEMBER_SELECT)) {
```
[the remark this answers is not in the document]
Comment (mpaladin) [reply java-frontend/src/main/java/org/sonar/java/se/checks/ObjectOutputStreamCheck.java]: :+1:

**Pull request merged**